client/modules/cm-lang-ftml/src/grammars/ftml.yaml

```diff
@@ -0,0 +1,999 @@
+    end: { match: /$/ }
+    inside:
+      - include: inline
+      - include: blocks
+
+  Table:
+    fold: past_first_line
+    begin:
+      match: /@sws\|{2}/
+      rematch: true
+    end:
+      lookbehind: '!/@esc\s*/'
+      match: /@enl|^(?:(?!\|{2}).)+$/
+      rematch: true
+    inside:
+      - type: TableMark
+        match: /(\|{2,})([<=~>]?)/
+        captures:
+          0: { type: TableMarkSeparator, tag: separator }
+          1: { type: TableMarkModifier, tag: modifier }
+      - include: inline
+      - include: blocks
+
+  ContainerMark:
+    match: /@sws>+|[*#]/
+    tag: keyword
+
+  Container:
+    fold: past_first_line
+    begin:
+      match: /@cs/
+      rematch: true
+    end:
+      lookbehind: '!/@esc\s*/'
+      match: /@enl|(?!@cs)@sws/
+      rematch: true
+    inside:
+      # simplified block markup inside containers
+      # horizontal rule
+      - match: /(@cs)(@hr)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: HorizontalRule }
+      # clearfloat
+      - match: /(@cs)(~{3,}[<>]?)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: ClearFloat }
+      # heading
+      - match: /(@cs)(@heading)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: HeadingMark }
+      # center
+      - match: /(@cs)(=@ws+)/
+        captures:
+          0: { is: ContainerMark }
+          1: { is: CenterMark }
+      # everything else
+      - ContainerMark
+      - include: inline
+      - include: blocks
+
+  # disabled for now
+  Paragraph:
+    begin: { match: /@sws(?!@interrupt)\S/, rematch: true }
+    end: { match: /@sws(?:@interrupt)|@enl/, rematch: true }
+    fold: past_first_line
+    inside:
+      - include: inline
+      - include: blocks
+
+  # -- TRIPLE LINK
+
+  TripleLinkStart:
+    match: /\[{3}(?!\[)/
+    tag: squareBracket
+    closedBy: TripleLinkEnd
+
+  TripleLinkEnd:
+    match: /(?!\]{4})\]{3}/
+    tag: squareBracket
+    openedBy: TripleLinkStart
+
+  TripleLinkSeparator:
+    match: "|"
+    tag: separator
+
+  TripleLinkURL:
+    match: /([*#])?([^*\n|\]]+)/
+    captures:
+      0: { type: TripleLinkURLModifier, tag: modifier }
+      1: { type: TripleLinkURLText, tag: link }
+
+  TripleLinkText:
+    match: /[^\n\]]+/
+
+  TripleLink:
+    skip: /@ws/
+    chain:
+      - TripleLinkStart
+      - TripleLinkURL
+      - TripleLinkSeparator
+      - TripleLinkText?
+      - TripleLinkEnd
+
+  TripleLinkSimple:
+    emit: TripleLink
+    skip: /@ws/
+    chain:
+      - TripleLinkStart
+      - TripleLinkURL
+      - TripleLinkEnd
+
+  # -- SINGLE LINK
+
+  SingleLinkStart:
+    match: /\[/
+    tag: squareBracket
+    closedBy: SingleLinkEnd
+
+  SingleLinkEnd:
+    match: /\]/
+    tag: squareBracket
+    openedBy: SingleLinkStart
+
+  SingleLinkSeparator:
+    match: "|"
+    tag: separator
+
+  SingleLinkSlug:
+    match: /([:#*]|(?=\/)|(?=[^#*\s]+?[@:][^#*\s]))([^#*\s]+)/
```

Comment:
I think this one is complicated enough to deserve a comment.

Reply:
oh yeah lmao this is the crazy regex for dumb links like `wikipedia:foo-bar` that wikidot supports

I'll add a comment

Reply:
Yeah... Real nonsense.